src/Streamly/Internal/Data/Unfold/Types.hs

```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
 instance Functor m => Functor (Unfold m a) where
     {-# INLINE fmap #-}
     fmap = map
+
+------------------------------------------------------------------------------
+-- Applicative
+------------------------------------------------------------------------------
+
+{-# INLINE const #-}
+const :: Applicative m => m b -> Unfold m a b
+const m = Unfold step inject
+
+    where
+
+    inject _ = pure False
+
+    step False = (`Yield` True) <$> m
+    step True = pure Stop
+
+-- | Outer product discarding the first element.
+--
+-- /Unimplemented/
+--
+{-# INLINE_NORMAL apSequence #-}
+apSequence :: -- Monad m =>
+    Unfold m a b -> Unfold m a c -> Unfold m a c
+apSequence (Unfold _step1 _inject1) (Unfold _step2 _inject2) = undefined
+
+-- | Outer product discarding the second element.
+--
+-- /Unimplemented/
+--
+{-# INLINE_NORMAL apDiscardSnd #-}
+apDiscardSnd :: -- Monad m =>
+    Unfold m a b -> Unfold m a c -> Unfold m a b
+apDiscardSnd (Unfold _step1 _inject1) (Unfold _step2 _inject2) = undefined
+
+data Cross s1 s2 sy x y = CrossOuter s1 y | CrossInner s1 sy s2 x
+
+-- | Create a cross product (vector product or cartesian product) of the
+-- output streams of two unfolds.
+--
+{-# INLINE_NORMAL cross #-}
+cross :: Monad m => Unfold m a b -> Unfold m a c -> Unfold m a (b, c)
+cross (Unfold step1 inject1) (Unfold step2 inject2) = Unfold step inject
+
+    where
+
+    inject x = do
+        s1 <- inject1 x
+        return $ CrossOuter s1 x
+
+    {-# INLINE_LATE step #-}
+    step (CrossOuter st1 sy) = do
+        r <- step1 st1
+        case r of
+            Yield x s -> do
+                s2 <- inject2 sy
+                return $ Skip (CrossInner s sy s2 x)
```

Comment:
You can probably inline `CrossInner`. Saves 1 indirection I guess.

Reply:
`CrossInner` is a loop, its not a one shot thing.